source/extensions/common/wasm/wasm_vm.cc

```diff
@@ -0,0 +1,31 @@
+#include "extensions/common/wasm/wasm_vm.h"
+
+#include <memory>
+
+#include "extensions/common/wasm/null/null.h"
+#include "extensions/common/wasm/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+thread_local Envoy::Extensions::Common::Wasm::Context* current_context_ = nullptr;
```

Comment:
What is the purpose of these thread_locals? Should this be using Envoy TLS?

Reply:
The WASM VMs are per-thread and when we call into them and then they call out again we need a tamper-proof way to know who made the call into the VM (we don't have any other context information and we can't trust the managed code in the VM to hand us that context.  This thread_local stores that context information.  Envoy TLS (AFAICT) is a mechanism for initializing, getting and setting cross-thread data in a consistent way across all workers in Envoy.   However, this piece of information is only used during a call into a WASM VM, so it is more like a dynamic variable in the Lisp sense (dynamically scoped global).  Consequently the c++11 standard thread_local declaration is a better match IMO.  This variable doesn't need to have a consistent value outside of the time during calls into the VM.